### PR TITLE
WIP Queue class prototype

### PR DESCRIPTION
**Pull request by noldor**
https://github.com/bcit-ci/CodeIgniter4/issues/117
I try to make a queue prototype.
If this fits what you want, I will make the datails.

### CODE REVIEW

**Comment by noldor** [comment]
usage:
```
<?php namespace App\Controllers;

class QueueTest extends \CodeIgniter\Controller
{
    public function enqueue()
    {
        $queue = service('queue');
        $queue->send('hoge');
    }

    public function fetch()
    {
        $queue = service('queue');
        $queue->fetch(
            function($data)
            {
                var_dump($data);
            }
        );
    }

    public function recieve()
    {
        $queue = service('queue');
        $queue->recieve(
            function($data)
            {
                var_dump($data);
            }
        );
    }
}
```

File: system/Queue/Handlers/DatabaseHandler.php [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	/*
+	public function setup()
+	{
+		CREATE TABLE ci_queue (
+			id          int(11) NOT NULL AUTO_INCREMENT,
+			queue_name  varchar(255) NOT NULL,
+			status      int(11) NOT NULL,
+			weight      int(11) NOT NULL,
+			retry_count int(11) NOT NULL DEFAULT 0,
+			exec_after  datetime NOT NULL,
```
Comment (lonnieezell) [inline system/Queue/Handlers/DatabaseHandler.php]: I see you have a exec_after command. This means we'd need a way to set that, like a `delay()` method or similar?

File: system/Queue/Handlers/DatabaseHandler.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	/**
+	 * constructor.
+	 *
+	 * @param  array $group_config
+	 * @param  \Codeigniter\Config\Queue $config
+	 */
+	public function __construct($group_config, \Codeigniter\Config\Queue $config)
+	{
+		$this->group_config = $group_config;
+		$this->config       = clone $config;
```
Comment (lonnieezell) [inline system/Queue/Handlers/DatabaseHandler.php]: I'm probably missing something here, but why clone the config? Do they get modified somewhere?

File: system/Queue/Handlers/DatabaseHandler.php [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+	/**
+	 * Recieve message from queueing system.
+	 * When there are no message, this method will wait.
+	 *
+	 * @param  callable $callback
+	 * @param  string   $queueName
+	 * @return boolean  whether callback is done or not.
+	 */
+	public function recieve(callable $callback, string $queueName = '') : bool
```
Comment (lonnieezell) [inline system/Queue/Handlers/DatabaseHandler.php]: `recieve` is spelled wrong. The i/e should be swapped: `receive()`

**Comment by noldor** [comment]
Thank you very much.
I will try write docs. But I'm not good at English (it's harder than coding for me), so please give me some time.

**Comment by lonnieezell** [comment]
It doesn't have to be great docs. I can help with that later. Just trying to picture how you intend it to be used.

**Comment by noldor** [comment]
I wrote functions I think the queue interface should have. Could you check it?

----
Queue
=====

overview
--------

This queue interface will be aware of RabbitMQ. There are 4 components.

 - Provider
 - Exchanger
 - Queue
 - Consumer

There are multiple Consumers listening to same Queue, however, each message is processed by just one Consumer. It will not process twice.

If error raise, message will get back to same queue and will be retried. Retry count is limited. If retry count is over, the message will be deleted.

retry delay
-----------

Sometimes, Consumer job fails repeatedly (e.g. when the mail server is temporary unavailable, etc...).
So we should retry with delay.
Delay time can be modified in the config file.

retry count limit
-----------------

When a job that always fails retry forever, the queue is full of such jobs.
So a job should be restricted the number of retrying.
Retry limit can be modified in the config file.

Retry count is passed to the callback function.

rejecting a job by Consumer
---------------------------

When a error arises, job will be retried automatically. Also we can retry a job manually.
If a specific exception is thrown, then this queue system will retry as failure.

endless loop on Consumer's job
------------------------------

When a job fall into a endless loop, it seems that the job is processing forever from the point of view of the Queue. The job will be neither success nor failure.
This queue system doesn't take care of it. A programmer should set `max_execution_time` or `set_time_limit()` properly.

On the Database Handler nobody can decide whether a job is processing or not, so the handler regards a job taking more than 300 sec as failure and retry it.

return without waiting after fetching a message
------------------------------------------------

Normally, a request via a browser should response ASAP. So `fetch()` returns right away whether there are messages or not.
`fetch()` work with a single message.

waiting when there are no messages
-----------------------------------

If a job works as background job, the job should wait for a messages when there are no messages.
Because starting PHP process require CPU power.

`receive()` is the same as `fetch()`, except it will wait if there are no messages.
`receive()` work with a single messages and we should not call it repeatly with endless loop like `while(1)`. PHP process will kill itself by `max_execution_time`. We should use `supervisor`, `daemontools`, `systemd` or the like. At worst, `crond` (on cheap hosting services).


----
Thank you for taking time.

**Comment by noldor** [comment]
Some functions are not implemented. I will make it from now on.

File: system/Queue/Handlers/DatabaseHandler.php (hunk 3 above)
Comment (noldor) [inline system/Queue/Handlers/DatabaseHandler.php]: oops... thank you for pointing it out.

**Comment by lonnieezell** [comment]
@noldor Those features look like great additions.

**Comment by lonnieezell** [comment]
@noldor are you still working on this? Looks like there were a couple of tests on the PostgreSQL front that didn't pass the last round of changes.

**Comment by noldor** [comment]
@lonnieezell 
I'm sorry for not contacting you for a long time.
I was busy in March, and bit busy in April too. I will just remove the error this weekend.
The first week of May is a large holiday in Japan, so I'd like to work with File Handler at that time. I'm sorry my work is late.

**Comment by lonnieezell** [comment]
@noldor No worries. Just checking to see if you were still interested in working on this, and it sounds like you are. Good to hear! And I completely understand being busy. I've had a bit of that myself a month ago where I didn't get to touch it. 

Thanks for responding, and looking forward to getting to play with this a bit !

**Comment by lonnieezell** [comment]
@noldor Any luck on this?

**Comment by jim-parry** [comment]
not to be too picky, but we expect commits to be GPG-signed :)

**Comment by jim-parry** [comment]
Replaced by #1797

**Pull request closed without merging**